### PR TITLE
[53540][feature] Home screen

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 53541: CreateUI1](https://edu-redmine.sun-asterisk.vn/issues/53541)
- [#Task 53560: CreateUI2](https://edu-redmine.sun-asterisk.vn/issues/53560)
- [#Task 53562: Fetch and display data](https://edu-redmine.sun-asterisk.vn/issues/53562)
## WHAT
- Create ui of home screen
- Display data
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

**Review by minhnn-1279 — APPROVED** [review]

File: app/src/main/java/com/example/themeal/data/source/MealDataSource.kt [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+import com.example.themeal.data.model.CategoryResponse
+import com.example.themeal.data.model.MealCollapse
+import com.example.themeal.data.model.MealResponse
+import com.example.themeal.data.source.api.MealService
+import com.example.themeal.data.source.database.MealDatabase
+
+class MealDataSource(
+    private val apiService: MealService,
+    private val mealDatabase: MealDatabase
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/data/source/MealDataSource.kt]: e truyền apiService vào đây thì local phải truyền DAO chứ

File: app/src/main/java/com/example/themeal/ui/LaunchActivity.kt [hunk 2]
```diff
@@ -6,10 +6,21 @@ import android.os.Handler
...
+
+    private val homeViewModel by inject<HomeViewModel>()
+
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
+        homeViewModel.apply {
+            getMealByFirstLetter()
+            getCategories()
+            getNewMeal()
+            getListMealRecent()
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/LaunchActivity.kt]: cho vào init{} của viewModel

File: app/src/main/java/com/example/themeal/ui/MainActivity.kt [hunk 3]
```diff
@@ -1,13 +1,43 @@
...
 
-class MainActivity : AppCompatActivity() {
+class MainActivity : BaseActivity<ActivityMainBinding>(ActivityMainBinding::inflate) {
+
+    private val listFragmentId = listOf(R.id.item_home, R.id.item_ingredient, R.id.item_favorite)
+    private val callbackViewPager = object : OnPageChangeCallback() {
+        override fun onPageSelected(position: Int) {
+            super.onPageSelected(position)
+            binding.bottomNavView.menu.findItem(listFragmentId[position]).isChecked = true
+        }
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/MainActivity.kt]: bỏ đi nhé

File: app/src/main/java/com/example/themeal/ui/home/HomeViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import com.example.themeal.data.model.Category
+import com.example.themeal.data.model.MealCollapse
+import com.example.themeal.data.repository.Repository
+
+class HomeViewModel(private val mealRepository: Repository.MealRepository) : BaseViewModel() {
+
+    val allMealList = MutableLiveData<List<MealCollapse>>()
+    val recentMealList = MutableLiveData<List<MealCollapse>>()
+    val categoryList = MutableLiveData<List<Category>>()
+    val newMealList = MutableLiveData<List<MealCollapse>>()
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/home/HomeViewModel.kt]: xem lại cách khai báo live data của Trường nhé

File: app/src/main/java/com/example/themeal/ui/home/adapter/CategoryAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    class ViewHolder(binding: ItemCategoryBinding) :
+        BaseViewHolder<Category, ItemCategoryBinding>(binding) {
+
+        override fun onBindData(data: Category) {
+            binding.textCategoryTitle.text = data.name
+            Glide.with(binding.root.context)
+                .load(data.thumbnailLink)
+                .placeholder(R.mipmap.ic_launcher)
+                .error(R.mipmap.ic_launcher)
+                .into(binding.imageCategory)
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/home/adapter/CategoryAdapter.kt]: extension

File: app/src/main/java/com/example/themeal/ui/home/adapter/HomeAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
+        if (holder is CategoryViewHolder) {
+            holder.onBindData(getItem(position))
+        } else {
+            if (getItem(position).listItem.isEmpty()) {
+                (holder as? MealViewHolder)?.onBindEmptyData(getItem(position))
+            } else {
+                (holder as? MealViewHolder)?.onBindData(getItem(position))
+            }
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/home/adapter/HomeAdapter.kt]: viết vào trong baseAdapter , adapter nào cũng cần bind view mà

File: app/src/main/java/com/example/themeal/ui/home/adapter/MealAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    class ViewHolder(binding: ItemMealBinding) :
+        BaseViewHolder<MealCollapse, ItemMealBinding>(binding) {
+
+        override fun onBindData(data: MealCollapse) {
+            binding.textTitleMeal.text = data.name
+            Glide.with(binding.root.context)
+                .load(data.getLinkPreview())
+                .placeholder(R.mipmap.ic_launcher)
+                .error(R.mipmap.ic_launcher)
+                .into(binding.imageMeal)
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/home/adapter/MealAdapter.kt]: .

File: app/src/main/java/com/example/themeal/ui/home/diffutil/HomeItemDiff.kt [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package com.example.themeal.ui.home.diffutil
+
+import androidx.recyclerview.widget.DiffUtil
+import com.example.themeal.data.model.HomeItem
+
+class HomeItemDiff : DiffUtil.ItemCallback<HomeItem>() {
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/themeal/ui/home/diffutil/HomeItemDiff.kt]: DiffUtil viết cùng với object cho anh

**Pull request merged**